src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs

```diff
@@ -28,16 +28,6 @@ internal ModelOperationsCatalog(IHostEnvironment env)
             Explainability = new ExplainabilityTransforms(this);
         }
 
-        public abstract class SubCatalogBase
```

Comment:
I may do more work to get rid of this sort of thing. These mysterious subclasses mean nothing to the user, and as part of the "central" object `MLContext`, we should probably choose some alternate mechanism. (I got rid of it here, but there's similar code elsewhere.)

Reply:
I like this change.  :+1:

---
In reply to: [262299101](https://github.com/dotnet/machinelearning/pull/2842#discussion_r262299101) [](ancestors = 262299101)

Reply:
Very good, I have done as I suggested, and got rid of the rest. We typically hide infrastructure behind this using the tricks of explicit implementations of internal interfaces, so I used the same in introducing a new interface `IInternalCatalog` (not the best name, perhaps, but as it is internal we are free to improve it later). This also allowed me to simplify `CatalogUtils` considerably.